### PR TITLE
Add role to upgrade postgresql on EL8

### PR DESCRIPTION
**Pull request by ehelms**

### CODE REVIEW

**Comment by ehelms** [comment]
@evgeni I think CI here is failing hard due to some changes that I, from looking in FAM, think you have overcome. When I tried to mimic some of those requirements and Makefile updates here I kept making a mess of things. Could you maybe give things a once over for any changes you know we need to make?

**Comment by evgeni** [comment]
@ehelms the `lint` issue should be fixed in https://github.com/theforeman/foreman-operations-collection/pull/23, but I can't wrap my head around how to fix `molecule`.

**Comment by ehelms** [comment]
Includes https://github.com/theforeman/foreman-operations-collection/pull/23 with some additional fixes I read about. Let me know thoughts on how I tackled it. There is some repetition that can be cleaned up later.

**Comment by ehelms** [comment]
I debated if I should drop the lint job given molecule does linting.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -119,7 +121,6 @@ jobs:
         ansible:
           - stable-2.9
           - stable-2.10
-          - devel
```
Comment (evgeni) [inline .github/workflows/main.yml]: I guess this is temporally?

File: roles/postgresql_upgrade/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+theforeman.operations.postgresql_upgrade
+========================================
+
+Upgrades Postgresql for a Foreman install.
```
Comment (evgeni) [inline roles/postgresql_upgrade/README.md]:
```suggestion
Upgrades PostgreSQL for a Foreman install.
```

File: roles/postgresql_upgrade/molecule/redhat/prepare.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+      yum:
+        name: https://yum.theforeman.org/releases/latest/el{{ ansible_distribution_major_version }}/x86_64/foreman-release.rpm
+        disable_gpg_check: true
+        state: present
+
+    - name: 'Setup Puppet 6 Repository'
+      yum:
+        name: https://yum.puppet.com/puppet6-release-el-{{ ansible_distribution_major_version }}.noarch.rpm
+        disable_gpg_check: true
+        state: present
```
Comment (evgeni) [inline roles/postgresql_upgrade/molecule/redhat/prepare.yml]: we have roles for that now ;)

File: roles/postgresql_upgrade/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
+---
+- name: Get version of postgres data directory
+  slurp:
+    path: /var/lib/pgsql/data/PG_VERSION
+  register: pg_version
+
+- when:
+    - ansible_distribution_major_version == '8'
```
Comment (evgeni) [inline roles/postgresql_upgrade/tasks/main.yml]: This should also check for this being *RHEL* 8, not Debian 8 (even if we don't support Debian 8 anymore)

Plus I'd put this check as the first task, `/var/lib/pgsql` won't exist on EL7 installs with SCL'ed PostgreSQL anymore.

File: roles/postgresql_upgrade/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      set_fact:
+        collate: "{{ postgres_output.stdout.split('|')[3] | trim }}"
+        ctype: "{{ postgres_output.stdout.split('|')[4] | trim }}"
+
+    - name: Stop postgresql
+      service:
+        name: postgresql
+        state: stopped
+
+    - name: Update postgresq.conf data directory
```
Comment (evgeni) [inline roles/postgresql_upgrade/tasks/main.yml:38]:
```suggestion
    - name: Update postgresql.conf data directory
```

Update to what? The task essentially deletes the config line.

File: roles/postgresql_upgrade/README.md (hunk 2 above)
Comment (evgeni) [reply roles/postgresql_upgrade/README.md]: I think this also should point out that this is only for EL8 systems with module-based PSQL, which will update 10 to 12, not any arbitrary versions.

File: .github/workflows/main.yml (hunk 1 above)
Comment (ehelms) [reply .github/workflows/main.yml]: Aye.

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -102,7 +102,9 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-pip-
       - name: Install dependencies
-        run: pip install --upgrade -r requirements-lint.txt
+        run: |
+          pip install ansible
```
Comment (ekohl) [inline .github/workflows/main.yml]: Why is it needed to install ansible separate?

**Comment by ehelms** [comment]
12 runs later it is green again

File: .github/workflows/main.yml (hunk 6 above)
Comment (evgeni) [reply .github/workflows/main.yml]: ansible-lint doesn't depend on it (anymore), but putting it into lint.txt would pull it in other places where we want the git checkout of ansible instead.

File: roles/foreman_repositories/tasks/redhat.yml [hunk 7]
```diff
@@ -23,4 +23,5 @@
     name: https://fedorapeople.org/groups/katello/releases/yum/{{ foreman_repositories_katello_version }}/katello/el{{ ansible_distribution_major_version }}/x86_64/katello-repos-latest.rpm
     disable_gpg_check: true
     state: present
+  retries: 3
```
Comment (evgeni) [inline roles/foreman_repositories/tasks/redhat.yml]: it feels wrong to put in a workaround for our CI into the role itself, not that I'd have a better solution :(

File: roles/postgresql_upgrade/README.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
+theforeman.operations.postgresql_upgrade
+========================================
+
+Upgrades PostgreSQL for a Foreman install on EL8 only.
```
Comment (evgeni) [inline roles/postgresql_upgrade/README.md]:
```suggestion
Upgrade Foreman's PostgreSQL from 10 to 12 on EL8.
```

File: roles/postgresql_upgrade/README.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+Role Variables
+--------------
+
```
Comment (evgeni) [inline roles/postgresql_upgrade/README.md]:
```suggestion
```

File: roles/postgresql_upgrade/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,62 @@
+---
+- name: Get version of postgres data directory
```
Comment (evgeni) [inline roles/postgresql_upgrade/tasks/main.yml]: I thought more about the following:

```yaml
- name: Ensure running on EL8
  assert:
    that:
      - ansible_os_family == "RedHat"
      - ansible_distribution_major_version == '8'
    fail_msg: "This role is only supported on EL8"

- name: Get version from the PostgreSQL data directory
  slurp:
    path: /var/lib/pgsql/data/PG_VERSION
  register: pg_version

- name: Ensure PostgreSQL version is 10
  assert:
    that:
      - "'10' in pg_version['content']|b64decode"
    fail_msg: "This role only supports upgrades from PostgreSQL 10"

[ rest of the tasks without the block ]
```

But wouldn't block on my personal preferences ;)

File: roles/postgresql_upgrade/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    - ansible_distribution_major_version == '8'
+    - "'10' in pg_version['content']|b64decode"
+  block:
+    - name: Start postgresql
+      service:
+        name: postgresql
+        state: started
+
+    - name: Collect postgres data
+      shell: psql -lt | grep -E "^\s+postgres"
```
Comment (evgeni) [inline roles/postgresql_upgrade/tasks/main.yml]:
```suggestion
      shell: psql --list --tuples-only | grep -E "^\s+postgres"
```

File: roles/postgresql_upgrade/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        state: present
+
+    - name: Update postgresql packages  # noqa package-latest
+      dnf:
+        name:
+          - postgresql-server
+          - postgresql-upgrade
+        state: latest
+
+    - name: Run pg_upgrade
```
Comment (evgeni) [inline roles/postgresql_upgrade/tasks/main.yml]:
```suggestion
    - name: Run postgresql-setup --upgrade
```

File: roles/postgresql_upgrade/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+        name: postgresql
+        state: stopped
+
+    - name: Remove data directory entry from postgresql.conf
+      lineinfile:
+        path: /var/lib/pgsql/data/postgresql.conf
+        regexp: '^data_directory ='
+        line: ''
+
+    - name: Reset postgresql module
```
Comment (ekohl) [inline roles/postgresql_upgrade/tasks/main.yml]: This doesn't look like a reset but rather an enable.

File: roles/postgresql_upgrade/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+      dnf:
+        name: '@postgresql:12'
+        state: present
+
+    - name: Update postgresql packages  # noqa package-latest
```
Comment (ekohl) [inline roles/postgresql_upgrade/tasks/main.yml:49]: This is odd. For me a `dnf module install postgresql:12` also upgraded all packages to the new version. Then it was sufficient for me to ensure `postgresql-update` was installed.

**Comment by evgeni** [comment]
I left a few nitpicky comments. I think the only thing worth debating is the addition of "retries" to the repositories role, which seems wrong to me, but gets the CI to a working setup…

**Review by ekohl — COMMENTED** [review]
Should the source and target version be variables? This seems like a good practice to me. My programming professor would fail us if we had a number other than 0, 1 or 2 in the code. They all had to be constants. It's something I still feel quite strongly about and it pains me to see `12` in multiple places.

File: roles/foreman_repositories/tasks/redhat.yml (hunk 7 above)
Comment (ehelms) [reply roles/foreman_repositories/tasks/redhat.yml]: I was hoping you would, cause re-running the job 12 times yesterday felt bad and this felt just enough less bad to be worth it.

One solution is to move Katello repos off of fedorapeople :P

File: roles/postgresql_upgrade/tasks/main.yml (hunk 10 above)
Comment (ehelms) [reply roles/postgresql_upgrade/tasks/main.yml]: The initial assert makes sense to me. The last assert will break idempotency I think?

Comment (evgeni) [reply roles/postgresql_upgrade/tasks/main.yml]: True.

Comment (evgeni) [reply roles/postgresql_upgrade/tasks/main.yml]: True.

File: .github/workflows/main.yml (hunk 6 above)
Comment (ekohl) [reply .github/workflows/main.yml]: The correct way to deal with is to call `pip install --upgrade -r requirements-lint.txt ansible`, not 2 separate calls to pip. This allows the pip's dependency resolution to work correctly.

**Comment by ehelms** [comment]
Rebased on the latest repository changes

**Comment by evgeni** [comment]
Seems this push here:
```
 @ehelms ehelms force-pushed the ehelms:postgres-upgrade branch from 8ee7e6e to dd33362 5 hours ago 
```

has overwritten a lot of changes we discussed and you did prior to the Katello repo fix :(

**Comment by ehelms** [comment]
I think I recovered all the updates I lost through the comments.

File: roles/postgresql_upgrade/tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+      - ansible_distribution_major_version == '8'
+    fail_msg: "This role is only supported on EL8"
+
+- name: Get version from the PostgreSQL data directory
+  slurp:
+    path: /var/lib/pgsql/data/PG_VERSION
+  register: pg_version
+  when:
+    - ansible_os_family == "RedHat"
+    - ansible_distribution_major_version == '8'
```
Comment (evgeni) [inline roles/postgresql_upgrade/tasks/main.yml]:
```suggestion
```

The `assert` above will already have bailed out if that's not the case

File: roles/postgresql_upgrade/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+- when:
+    - ansible_os_family == "RedHat"
+    - ansible_distribution_major_version == '8'
```
Comment (evgeni) [inline roles/postgresql_upgrade/tasks/main.yml]:
```suggestion
```

The `assert` above will already have bailed out if that's not the case

Comment (ehelms) [reply roles/postgresql_upgrade/tasks/main.yml]: This doesn't hurt though :)

File: roles/postgresql_upgrade/tasks/main.yml (hunk 15 above)
Comment (ehelms) [reply roles/postgresql_upgrade/tasks/main.yml]: This doesn't hurt though :)

File: roles/postgresql_upgrade/molecule/redhat/prepare.yml [hunk 17]
```diff
@@ -0,0 +1,10 @@
+---
+- name: Prepare
+  hosts: all
+  gather_facts: true
+  roles:
+    - puppet_repositories
+    - foreman_repositories
+    - role: installer
+      vars:
+        installer_scenario: foreman
```
Comment (evgeni) [inline roles/postgresql_upgrade/molecule/redhat/prepare.yml]: should this explicitly point at a foreman version that came with psql 10, as we plan to shift the default to 12 soon?

File: roles/postgresql_upgrade/tasks/main.yml (hunk 15 above)
Comment (evgeni) [reply roles/postgresql_upgrade/tasks/main.yml]: Sure, but my eyes/brain ;)

File: roles/postgresql_upgrade/molecule/redhat/prepare.yml (hunk 17 above)
Comment (ehelms) [reply roles/postgresql_upgrade/molecule/redhat/prepare.yml]: Good point, I'll point it at 2.3.

File: roles/postgresql_upgrade/tasks/main.yml (hunk 15 above)
Comment (ehelms) [reply roles/postgresql_upgrade/tasks/main.yml]: Better to be explicit and safe? Just never look at this code again ;)

**Review by evgeni — APPROVED** [review]
I'd *prefer* if the `when`s would be trimmed, but won't block on it.

**Pull request merged**